pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1772,21 +1772,51 @@ func (c *Cloud) DetachDisk(diskName KubernetesVolumeID, nodeName types.NodeName)
 		return "", err
 	}
 
-	awsInstance, info, err := c.getFullInstance(nodeName)
+	info, err := disk.describeVolume()
+
 	if err != nil {
-		if err == cloudprovider.InstanceNotFound {
-			// If instance no longer exists, safe to assume volume is not attached.
-			glog.Warningf(
-				"Instance %q does not exist. DetachDisk will assume disk %q is not attached to it.",
-				nodeName,
-				diskName)
+		glog.Errorf("Error describing volume %q: %q", disk.awsID, err)
+		// if for some reason we can not describe volume we will return error and assume disk is detached
+		// because it is probably deleted
+		return "", nil
+	}
+
+	volumeState := aws.StringValue(info.State)
+
+	// If volume is available it means it is detached
+	if volumeState == "available" {
+		return "", nil
+	}
+
+	var instanceInfo *ec2.Instance
+
+	if len(info.Attachments) > 0 {
+		attachment := info.Attachments[0]
+		instanceID := aws.StringValue(attachment.InstanceId)
+		instanceInfo, err = c.getInstanceByID(instanceID)
+
+		// This should never happen but if it does it could mean there was a race and instance
+		// has been deleted, in which case volume is detached anyways.
+		if err != nil {
+			glog.Errorf("Error fetching instance %s for volume %s", instanceID, diskName)
 			return "", nil
 		}
 
-		return "", err
+		attachedNodeName := mapInstanceToNodeName(instanceInfo)
+
+		if attachedNodeName != nodeName {
+			glog.Warningf("DetachDisk %s called for node %s but volume is attached to node %s", diskName, nodeName, attachedNodeName)
+			return "", nil
+		}
+	}
```

Comment:
is there an `else` err to return here where volumeState is not "available" and there are no attachments?